### ISSUE TITLE
Updating OracleJDK 8

### ISSUE THREAD

**Issue by FWest98** [travis-ci/travis-ci#3259]
Due to a bug in the JDK (8u31, https://bugs.openjdk.java.net/browse/JDK-8044546) I'm unable to build my Android app. I had to disable Travis for my repository.
As this problem is fixed in 8u40, which is planned for March, I was wondering how long it takes before Travis runs the new Java version.
Another question, isn't there a way to force a specific Java update in the Travis build?

**Comment by BanzaiMan** [comment]
We currently rely on https://launchpad.net/~webupd8team/+archive/ubuntu/java to provide Oracle JDK 8. There is work underway to bring OpenJDK 8, too, but I'm not sure when it'll have 8u40.

The next scheduled build environment update is early April.

If you can find a PPA to supply a JDK 8 package, you can use `sudo apt-get …` to get your environment set up for your builds.

**Comment by ssaring** [comment]
The repository https://launchpad.net/~webupd8team/+archive/ubuntu/java now provides Oracle Java SE 8u40 packages. It would be great when they could get installed on the Travis systems, my project depends on the new JavaFX API introduced in this Java version.

**Comment by zrcoder** [comment]
Now, you can use `linuxbrew` to install JDK 8_40.
I use the following script to build and test SBT application.

```
before_install:
    - rm -rf ~/.linuxbrew
    - yes | ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/linuxbrew/go/install)"
    - export PATH="$HOME/.linuxbrew/bin:$PATH"
    - export MANPATH="$HOME/.linuxbrew/share/man:$MANPATH"
    - export INFOPATH="$HOME/.linuxbrew/share/info:$INFOPATH"
    - brew install jdk
    - brew install scala
    - brew install sbt
script:
    - sbt test
```

**Comment by BanzaiMan** [comment]
The next update will include Java 8u40.

**Comment by davidmoten** [comment]
good to hear, my builds are breaking also for this reason. When is the next update?

**Comment by BanzaiMan** [comment]
We've had to postpone the April update. It will be announced in http://docs.travis-ci.com/user/build-environment-updates/2015-04-09/ and the mailing list (see http://docs.travis-ci.com/user/build-environment-updates/).

While we did that, 8u45 became available, so the update will contain that.

If you need 8u45 *now*, you can use the standard infrastructure and do something like:

```yaml
language: java

sudo: required

jdk:
  - oraclejdk8

install: true

script:
  - sudo apt-get update && sudo apt-get install oracle-java8-installer
  - java -version
```

I haven't tested this on non-java image; not clear if it works there.

**Comment by hrj** [comment]
Ping! Any updates?

The latest java release from oracle is 8u51. My builds are failing too because of the older version in travis.

**Comment by talios** [comment]
+1 on this. Just switched to Travis for a project and hitting this :(

**Comment by dobesv** [comment]
When I try the solution suggested by @BanzaiMan above, I get:

```
$ sudo apt-get update && sudo apt-get install oracle-java8-installer
sudo: must be setuid root
The command "sudo apt-get update && sudo apt-get install oracle-java8-installer" failed and exited with 1 during .
Your build has been stopped.
```

**Comment by dobesv** [comment]
I tried the linuxbrew receipt above, but it also errored:

```
$ yes | ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/linuxbrew/go/install)"
==> This script will install:
/home/travis/.linuxbrew/bin/brew
/home/travis/.linuxbrew/Library/...
/home/travis/.linuxbrew/share/man/man1/brew.1
==> Downloading and installing Homebrew...
remote: Counting objects: 279826, done.        
remote: Compressing objects: 100% (5/5), done.        
remote: Total 279826 (delta 0), reused 0 (delta 0), pack-reused 279821        
Receiving objects: 100% (279826/279826), 53.37 MiB | 21.23 MiB/s, done.
Resolving deltas: 100% (210443/210443), done.
From https://github.com/Homebrew/linuxbrew
 * [new branch]      master     -> origin/master
HEAD is now at 4321528 libuuid 1.0.3: Add tag linuxbrew
Warning: /home/travis/.linuxbrew/bin is not in your PATH.
==> Installation successful!
==> Next steps
Install the Linuxbrew dependencies:
Debian, Ubuntu, etc.:
  `sudo apt-get install build-essential`
Fedora, Red Hat, CentOS, etc.:
  `sudo yum groupinstall 'Development Tools'`
See http://brew.sh/linuxbrew/#dependencies for more information.
Run `brew doctor` before you install anything
Run `brew help` to get started
sudo: must be setuid root
$ export PATH="$HOME/.linuxbrew/bin:$PATH"
$ export MANPATH="$HOME/.linuxbrew/share/man:$MANPATH"
$ export INFOPATH="$HOME/.linuxbrew/share/info:$INFOPATH"
$ brew install jdk
sudo: must be setuid root
==> Downloading http://download.oracle.com/otn-pub/java/jdk/8u40-b26/jdk-8u40-li
==> Downloading from http://download.oracle.com/errors/download-fail-1505220.htm
######################################################################## 100.0%
Error: SHA256 mismatch
Expected: da1ad819ce7b7ec528264f831d88afaa5db34b7955e45422a7e380b1ead6b04d
Actual: 80212f7748c2671b89b6085000717c747851db004409dce3bb97f3a2aeb91cdd
Archive: /home/travis/.cache/Homebrew/jdk-1.8.0-40.tar.gz
To retry an incomplete download, remove the file above.
The command "brew install jdk" failed and exited with 1 during .
Your build has been stopped.
```

**Comment by dobesv** [comment]
The solution by @BanzaiMan worked for me after I removed `sudo: false` in my `.travis.yml`

**Comment by dobesv** [comment]
To make this work on the "new container based infrastructure" add this to your travis.yml:

```
sudo: false
addons:
  apt:
    packages:
      - oracle-java8-installer
```

**Comment by hrj** [comment]
:+1: @dobesv 

Thanks, that last comment worked for me!

**Comment by prohde** [comment]
I needed way to much commits to get my build passed. I would appreciate if at least JDK 8u40 would be installed.

**Comment by innovimax** [comment]
:+1:

**Comment by FibreFoX** [comment]
Any progress on this? Is there any ETA for this feature?

**Comment by hrj** [comment]
The [new ubuntu 14.04](http://blog.travis-ci.com/2015-10-14-opening-up-ubuntu-trusty-beta/) environment has Oracle JDK 8u51.

**Comment by fpinjava** [comment]
Travis is still building with 1.8.0_31 when the current version is now 1.8.0_66 !

**Comment by FibreFoX** [comment]
@fpinjava like suggested in the [comment above](https://github.com/travis-ci/travis-ci/issues/3259#issuecomment-150712876) the new "beta-"ubuntu does come with OracleJDK "1.8.0_51" and OpenJDK "1.8.0_45-internal", which is good, but does run slow (because they are not the "new container based infrastructure")

just change your travis-file by adding this:
```
sudo: required
dist: trusty
```

@hrj do you know any trick to get openjfx installed aswell? javafx is part of oracleJDK, but not on openjdk ... and thats what I need currently

**Comment by fpinjava** [comment]
@FibreFoX I need to check the build with the last version of the JDK, which I do by downloading it. The only problem is that it can't be done (at this time) in a container because sudo is not supported.

**Comment by jodastephen** [comment]
It is very unfortunate to see this issue languishing on Travis. JDK 8u31 is **very** old. It predates the critical 8u40 release when the worst of the lambda compilation bugs were fixed. For many projects, running on JDK 8 is not feasible - it has to be 8u40 or later.

While there are obviously workarounds, they slow down the build and put more strain on Travis' servers than necessary.

Given that it looks like there has been no update in a year now, it also leaves a question over the health of Travis more generally.

**Comment by xea** [comment]
Can we get an ETA on this, please?

**Comment by z3ntu** [comment]
Why is it still  1.8.0_31 ?

**Comment by BanzaiMan** [comment]
Sorry about the delay. We are working on the next update, which should fix this issue. There are a few housekeeping issues we need to take care of before we can roll it out.

We hope to have an announcement soon. I am locking this issue in the meantime.

**Comment by acnagy** [comment]
Hey everyone - thanks for your patience. Just wanted to update that the newer `sudo: required` GCE images now have OracleJDK 8. To answer @z3ntu's concern above, specifically, Trusty now has 1.8u101. The Docker image, however, is still a bit behind so @dobesv's [suggestion](https://github.com/travis-ci/travis-ci/issues/3259#issuecomment-130860338) is still good.

**Comment by BanzaiMan** [comment]
I think this can be closed now.
